File: src/net/azib/photos/RequestRouter.java
package net.azib.photos;

import org.apache.velocity.Template;
import org.apache.velocity.VelocityContext;
import org.apache.velocity.app.VelocityEngine;

import javax.servlet.*;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Enumeration;
import java.util.MissingResourceException;
import java.util.Properties;
import java.util.logging.Logger;

import static javax.servlet.http.HttpServletResponse.SC_MOVED_PERMANENTLY;
import static javax.servlet.http.HttpServletResponse.SC_NOT_FOUND;
import static javax.xml.bind.DatatypeConverter.parseInt;

public class RequestRouter implements Filter {
  private static final Logger logger = Logger.getLogger(RequestRouter.class.getName());
  private ServletContext context;
  private static VelocityEngine velocity;

  public void init(FilterConfig config) throws ServletException {
    this.context = config.getServletContext();
    Properties velocityProps = new Properties();
    velocityProps.setProperty("file.resource.loader.path", context.getRealPath("/WEB-INF/views"));
    velocityProps.setProperty("file.resource.loader.cache", "true");
    velocity = new VelocityEngine(velocityProps);
    velocity.setApplicationAttribute("javax.servlet.ServletContext", context);
    velocity.init();
  }

  public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain) throws IOException, ServletException {
    HttpServletRequest request = (HttpServletRequest) req;
    HttpServletResponse response = (HttpServletResponse) resp;
    String path = request.getServletPath();

    String by = request.getParameter("by");
    String random = request.getParameter("random");
    String userAgent = request.getHeader("User-Agent");
    boolean bot = isBot(userAgent);
    if (bot && (by != null || random != null)) {
      response.sendRedirect("/");
      response.setStatus(SC_MOVED_PERMANENTLY);
      return;
    }

    try {
      Picasa picasa = new Picasa(by, request.getParameter("authkey"));
      request.setAttribute("picasa", picasa);
      request.setAttribute("host", request.getHeader("host"));
      request.setAttribute("servletPath", request.getServletPath());
      request.setAttribute("bot", bot);
      request.setAttribute("mobile", userAgent != null && userAgent.contains("Mobile") && !userAgent.contains("iPad") && !userAgent.contains("Tab"));

      if (random != null) {
        request.setAttribute("delay", request.getParameter("delay"));
        if (request.getParameter("refresh") != null) request.setAttribute("refresh", true);
        render("random", picasa.getRandomPhotos(parseInt(random.length() > 0 ? random : "1")), request, response);
      }
      else if (path == null || "/".equals(path)) {
        if (request.getParameter("reload") != null)
          new CacheReloader().reload();

        render("gallery", picasa.getGallery(), request, response);
      }
      else if (path.lastIndexOf('.') >= path.length() - 4) {
        chain.doFilter(req, resp);
      }
      else {
        String[] parts = path.split("/");
        Album album;
        try {
          album = picasa.getAlbum(parts[1]);
        }
        catch (MissingResourceException e) {
          album = picasa.search(parts[1]);
          album.title = "Photos matching '" + parts[1] + "'";
          // TODO: no longer works for non-logged-in requests
        }

        if (parts.length > 2) {
          for (Photo photo : album.photos) {
            if (photo.id.equals(parts[2])) {
              request.setAttribute("photo", photo);
              break;
            }
          }
        }
        render("album", album, request, response);
      }
    }
    catch (MissingResourceException e) {
      response.sendError(SC_NOT_FOUND);
    }
  }

  static boolean isBot(String userAgent) {
    return userAgent == null || userAgent.toLowerCase().contains("bot/") || userAgent.contains("spider/");
  }

  static void render(String template, Object source, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
    long start = System.currentTimeMillis();

    request.setAttribute(template, source);

    if (response.getContentType() == null)
      response.setContentType("text/html; charset=utf8");
    if (source instanceof Entity && ((Entity) source).timestamp != null)
      response.addDateHeader("Last-Modified", ((Entity) source).timestamp);

    VelocityContext ctx = new VelocityContext();
    Enumeration<String> attrs = request.getAttributeNames();
    while (attrs.hasMoreElements()) {
      String name =  attrs.nextElement();
      ctx.put(name, request.getAttribute(name));
    }
    Template tmpl = velocity.getTemplate(template + ".vm");
    tmpl.merge(ctx, response.getWriter());

    logger.info("Rendered in " + (System.currentTimeMillis() - start) + " ms");
  }

  public void destroy() {
  }
}

File: src/net/azib/photos/RequestRouter.kt
package net.azib.photos

import java.util.*
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import javax.servlet.http.HttpServletResponse.SC_MOVED_PERMANENTLY
import javax.servlet.http.HttpServletResponse.SC_NOT_FOUND

class RequestRouter(val req: HttpServletRequest, val res: HttpServletResponse, val render: Renderer, val contentLoader: ContentLoader, val chain: FilterChain) {
  val attrs: MutableMap<String, Any?> = HashMap()
  val userAgent: String? = req.getHeader("User-Agent")
  val path = req.servletPath
  val pathParts = path.substring(1).split("/")
  var requestedUser = req["by"]
  val random = req["random"]
  val searchQuery = req["q"]
  var picasa = Picasa(requestedUser, req["authkey"])
  var bot = false

  fun invoke() {
    try {
      detectMobile()
      detectBot()

      attrs["picasa"] = picasa
      attrs["host"] = req.getHeader("host")
      attrs["servletPath"] = path

      if (req["reload"] != null) URLLoader.reload()

      when {
        random != null -> renderRandom()
        searchQuery != null -> renderSearch(searchQuery)
        (path == null || "/" == path) && requestedUser == null -> throw Redirect(picasa.urlPrefix)
        picasa.urlPrefix == path || "/" == path -> renderGallery()
        path.isResource() -> chain.doFilter(req, res)
        // pathParts.size == 1 -> throw Redirect(picasa.urlPrefix + path)
        pathParts.size == 2 && pathParts[1].matches("\\d+".toRegex()) -> redirectOldPhotoUrl()
        else -> renderAlbum(pathParts.last())
      }
    }
    catch (e: Redirect) {
      res.status = SC_MOVED_PERMANENTLY
      res.setHeader("Location", e.path)
    }
    catch (e: MissingResourceException) {
      res.sendError(SC_NOT_FOUND)
    }
  }

  fun String.isResource() = lastIndexOf('.') >= length - 4

  private fun renderGallery() {
    render("gallery", picasa.gallery.addMissingAlbums(), attrs, res)
  }

  private fun Gallery.addMissingAlbums(): Gallery {
    val missingNames = contentLoader.albums.keys - albums.keys
    missingNames.forEach { name -> albums[name] = Album(name=name, content=contentLoader.albums[name]) }
    return this
  }

  private fun renderSearch(q: String) {
    // TODO: no longer works for non-logged-in requests
    val album = picasa.search(q)
    album.title = "Photos matching '$q'"
    render("album", album, attrs, res)
  }

  private fun redirectOldPhotoUrl() {
    val lastSlashPos = path.lastIndexOf('/')
    if (bot) throw MissingResourceException(path, "", "")
    else throw Redirect(path.replaceRange(lastSlashPos, lastSlashPos+1, picasa.urlSuffix + "#"))
  }

  private fun renderAlbum(name: String) {
    var album: Album
    try {
      album = picasa.getAlbum(name).addContent()
      if (album.id == name && album.id != album.name)
        throw Redirect("/${album.name}${picasa.urlSuffix}")
    }
    catch (e: MissingResourceException) {
      album = Album(title=pathParts[0], description="No such album", author=picasa.gallery.author)
      res.status = SC_NOT_FOUND
    }

    render("album", album, attrs, res)
  }

  private fun Album.addContent(): Album {
    content = contentLoader.albums[name]
    return this
  }

  private fun renderRandom() {
    attrs["delay"] = req["delay"]
    attrs["refresh"] = req["refresh"] != null
    val numRandom = (if (random.isNotEmpty()) random else "1").toInt()
    render("random", picasa.getRandomPhotos(numRandom), attrs, res)
  }

  private fun detectMobile() {
    attrs["mobile"] = userAgent != null && userAgent.contains("Mobile") && !userAgent.contains("iPad") && !userAgent.contains("Tab")
  }

  private fun detectBot() {
    bot = isBot(userAgent)
    if (bot && requestedUser != null) {
      throw Redirect("/${Picasa.defaultUser}")
    }
    attrs["bot"] = bot
  }

  internal fun isBot(userAgent: String?): Boolean {
    return userAgent == null || userAgent.contains("bot/", true) || userAgent.contains("spider/", true)
  }

  class Redirect(val path: String): Exception()
}

operator fun HttpServletRequest.get(param: String) = getParameter(param)
